### PR TITLE
Triton&Flashibi in examples/llm

### PR DESCRIPTION
**Pull request by vchiley**
This pr integrates the triton implementation of FlashAttn (with ALiBi) into examples/llm.
Using this variant, memory usage is the same
Screenshot 2022-12-22 at 4 54 33 PM
it runs slightly faster
Screenshot 2022-12-22 at 4 51 15 PM
and with Alibi convergence is faster (and we converge to a lower final loss)
Screenshot 2022-12-22 at 4 51 08 PM
Screenshot 2022-12-22 at 4 51 23 PM

Meta-comment: MosaicGPT is currently structured so that the Causal Attn Bias is a member of the layer. It should really be a member of the model and should then be passed to all layers in their fwd pass. FlashAttn does not take a Bias / Mask so it does not suffer the memory penalty of having multiple masks. The Torch and Triton implementations do therefore are affected by the issue. @abhi-mosaic maybe we should create an issue to track this. We could make it part of this PR, but its slightly outside the scope.

Note: I tried a symlink `bert/src/flash_attn_triton.py -> llm/src/flash_attn_triton.py` but the result was a copy pasta; I'm not sure git supports symlinks...

### CODE REVIEW

File: llm/requirements.txt [hunk 1]
```diff
@@ -4,6 +4,8 @@ torch==1.12.1
 mosaicml==0.11.1
 mosaicml-streaming==0.1.2
 flash-attn==0.2.2
+# need a newer version of triton
+triton==2.0.0.dev20221103
```
Comment (abhi-mosaic) [inline llm/requirements.txt:8]: any particular reason for this version vs. `2.0.0.dev20221202`

File: llm/src/mosaic_gpt.py [hunk 2]
```diff
@@ -149,20 +231,18 @@ def __init__(self, cfg: DictConfig):
...
-                                 device=cfg.device),
-                wpe=nn.Embedding(cfg.max_seq_len,
-                                 cfg.d_model,
-                                 device=cfg.device),
-                emb_drop=nn.Dropout(cfg.emb_pdrop),
-                blocks=nn.ModuleList([
+        
+        self.alibi = cfg.get('alibi', False)
+        
+        self.transformer = nn.ModuleDict({"wte": nn.Embedding(cfg.vocab_size, cfg.d_model, device=cfg.device)})
```
Comment (abhi-mosaic) [inline llm/src/mosaic_gpt.py:237]:
```suggestion
        self.transformer = nn.ModuleDict({'wte': nn.Embedding(cfg.vocab_size, cfg.d_model, device=cfg.device)})
```

File: llm/src/mosaic_gpt.py [hunk 3]
```diff
@@ -178,7 +258,9 @@ def forward(self,
                            device=input_ids.device).unsqueeze(0)
 
         tok_emb = self.transformer.wte(input_ids)  # type: ignore
-        pos_emb = self.transformer.wpe(pos)  # type: ignore
+        pos_emb = 0
```
Comment (abhi-mosaic) [inline llm/src/mosaic_gpt.py:261]: does the adding 0 add any overhead? Idk if PyTorch is smart enough to skip

**Review by abhi-mosaic — APPROVED** [review]
LGTM! 

Acknowledged on the `attn_bias` buffer, lets refactor that in a followup PR.

Also, we should add CI to verify that certain files are the same across `examples/[llm,bert]` but this can come in the new year. Also, we may end up refactoring into a common/ folder and avoid this issue.

File: llm/requirements.txt (hunk 1 above)
Comment (vchiley) [reply llm/requirements.txt:8]: copy pasta [req from bert](https://github.com/mosaicml/examples/blob/main/bert/requirements.txt#L9)

**Comment by vchiley** [comment]
waiting for https://github.com/mosaicml/composer/pull/1854

File: llm/requirements.txt [hunk 4]
```diff
@@ -4,6 +4,8 @@ torch==1.13.0
 mosaicml==0.12.0
 mosaicml-streaming==0.1.2
 flash-attn==0.2.2
+# need a newer version of triton
+triton==2.0.0.dev20221103
```
Comment (abhi-mosaic) [inline llm/requirements.txt:8]:
```suggestion
triton==2.0.0.dev20221202
```

Any reason not to use the latest version?

File: llm/src/mosaic_gpt.py [hunk 5]
```diff
@@ -113,11 +122,15 @@ class GPTBlock(nn.Module):
...
     def __init__(self, cfg: DictConfig, device: Optional[str] = None):
         super().__init__()
+        if cfg.get('alibi', False):
+            assert 'triton' in cfg.attn_impl, 'Only triton kernel supports alibi'
         self.ln_1 = nn.LayerNorm(cfg.d_model, device=device)
         if cfg.attn_impl == 'torch':
             self.causal_attn = TorchCausalAttention(cfg, device)
         elif cfg.attn_impl == 'flash':
             self.causal_attn = FlashCausalAttention(cfg, device)
+        elif 'triton' in cfg.attn_impl:
```
Comment (abhi-mosaic) [inline llm/src/mosaic_gpt.py:132]: Could you elaborate why we're looking for `'triton' in cfg.attn_impl` rather than a fixed key like `cfg.attn_impl == 'triton'`?

File: llm/src/mosaic_gpt.py [hunk 6]
```diff
@@ -145,60 +159,167 @@ def __init__(self, cfg: DictConfig):
         super().__init__()
         assert cfg.name == 'mosaic_gpt', f'Tried to build MosaicGPT model with cfg.name={cfg.name}'
         self.cfg = cfg
+        self.alibi = cfg.get("alibi", None)
```
Comment (abhi-mosaic) [inline llm/src/mosaic_gpt.py:162]:
```suggestion
        self.alibi = cfg.get("alibi", False)
```

File: llm/src/mosaic_gpt.py [hunk 7]
```diff
@@ -145,60 +159,167 @@ def __init__(self, cfg: DictConfig):
...
+        #   mask to your attentions. See https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html#torch.nn.MultiheadAttention.forward
+        #   Basically True/-inf indicates tokens we do not want to attend to.
+        #
+        # 2. This is is the exact opposite behavior of Huggingface's tokenizers, which use the convention that True denotes tokens
+        #   we do want to attend to. See https://huggingface.co/docs/transformers/glossary#attention-mask
+        if self._attn_mask_initialized:
+            return self.attn_mask
+        
+        if not S:
+            S = x.size(1)
```
Comment (abhi-mosaic) [inline llm/src/mosaic_gpt.py:263]: Is `S` needed here?

File: llm/src/flash_attn_triton.py [hunk 8]
```diff
@@ -0,0 +1,1112 @@
+# Copyright 2022 MosaicML Examples authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Triton implementation of Flash Attention.
```
Comment (abhi-mosaic) [inline llm/src/flash_attn_triton.py:4]: Just want to double check on this file, is it a direct copy of https://github.com/HazyResearch/flash-attention/blob/main/flash_attn/flash_attn_triton.py?

And if so, is it up to date (last change was 2 weeks ago)?

File: llm/src/mosaic_gpt.py [hunk 9]
```diff
@@ -30,40 +30,16 @@ def __init__(self, cfg: DictConfig, device: Optional[str] = None):
...
-            self._fill_causal_attn_mask()
-            self.mask_initialized = True
-
-        return self.mhsa(x,
-                         x,
-                         x,
-                         attn_mask=self.mask,
-                         key_padding_mask=~key_padding_mask,
-                         need_weights=True)
+    def forward(self, x, key_padding_mask, attn_mask=None):
```
Comment (abhi-mosaic) [inline llm/src/mosaic_gpt.py:35]: Does this mean Torch-native attention now works with torch 1.13? I had this listed as a bug but it looks like you may have fixed it? 👀

**Review by abhi-mosaic — APPROVED** [review]
I made some minor comments, I think it would be good to have someone with more familiarity with ALIBI (@growlix , @alextrott16 )? to look over the attn_mask initialization. I wouldn't be able to tell if something is wrong.

The convergence improvement looks promising though! Maybe that's enough to support merging. Leaving it up to you @vchiley. 

Also, as we have added new features to `MosaicGPT`, the class has gotten quite big. I'll add a JIRA issue to refactor some of this, add comments, maybe move param_init_fn into a separate class, etc. It's getting harder to read and I want to make sure that when new users see the codebase for the first time, it's not too overwhelming.

File: llm/requirements.txt (hunk 4 above)
Comment (vchiley) [reply llm/requirements.txt:8]: I copied it from the [bert requirements](https://github.com/mosaicml/examples/blob/main/bert/requirements.txt#L8)

File: llm/src/flash_attn_triton.py (hunk 8 above)
Comment (vchiley) [reply llm/src/flash_attn_triton.py:4]: this is from the [bert dir](https://github.com/mosaicml/examples/blob/main/bert/src/flash_attn_triton.py)

I guess it points to the need for a unified code dir.

File: llm/src/mosaic_gpt.py (hunk 5 above)
Comment (vchiley) [reply llm/src/mosaic_gpt.py:132]: just me being dumb, I'll update it to follow convention.

File: llm/src/mosaic_gpt.py (hunk 7 above)
Comment (vchiley) [reply llm/src/mosaic_gpt.py:263]: nope

File: llm/src/mosaic_gpt.py (hunk 9 above)
Comment (vchiley) [reply llm/src/mosaic_gpt.py:35]: not sure where the bug was / wasn't but [this test](https://github.com/vchiley/examples/blob/tritonFlashibiFaster/llm/tests/test_hf_v_mosaic_gpt.py#L18) checks the hf model against mosaic_gpt with torch attn and it passes.

Also the mask generation (including the comments you made about the mask generation) I moved to [where I generate the mask](https://github.com/vchiley/examples/blob/tritonFlashibiFaster/llm/src/mosaic_gpt.py#L250).

**Comment by vchiley** [comment]
cc @abhi-mosaic 
> refactor

Probably a good idea if / when we create a 'common' folder which has (handles) attn, flash attn, triton flash attn, alibi etc

File: llm/src/flash_attn_triton.py (hunk 8 above)
Comment (alextrott16) [reply llm/src/flash_attn_triton.py:4]: @abhi-mosaic Yes, but we have not been incorporating changes to that original implementation into the BERT implementation.
Agreed that a location for unified code would be advantageous. I support the choice to re-use the BERT code for now, since it will be easier to refactor things if it's just a matter of reducing redundancy.

**Review by alextrott16 — APPROVED** [review]
Overall: I think alibi is correctly implemented, so I'm approving on that basis.

I'll add, though, that the general implementation of the 3 attention versions makes this a bit hard to read. I trust that there are good reasons to structure the code this way, but I think it warrants some detailed comments around the code in `MosaicGPT` where the attention mask is built. Also, in the attention classes themselves, it would help to add comments to explain how the associated methods will be used.

Last note: perhaps due to being a bit hard to parse, I'm not totally clear on how much re-building of the attention mask happens on each forward call. I'm assuming that re-building is minimal (e.g., just handling batch-specific masks due to padding). If that's not the case, I can provide a more fine-grained review on possible ways to re-use pre-computed masks.

**Comment by vchiley** [comment]
@alextrott16 
> how much re-building of the attention mask happens on each forward call

The generic and alibi attn masks are built and part of the model (`shape: 1 x 1 x 1 x SeqLen` or if alibi is used `shape: 1 x heads x 1 x SeqLen`; broadcasted to `shape: 1 x 1 x SeqLen x SeqLen` internally in the kernel; `causal=True` takes care of the causal portion of the mask)
These masks are only extended to `shape: BatchSize x 1 x 1 x SeqLen` when `key_padding_mask` says some of the tokens are invalid (only happens at eval since durring training `group_method: concat`).
As a result, yes "re-building is minimal", I "re-use pre-computed masks" as much as possible.

File: llm/src/flash_attention.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+                      (default: 1/sqrt(d_keys) where d_keys is computed at
+                      runtime)
+        attention_dropout: The dropout rate to apply to the attention
+                           (default: 0.0)
+    """
+    def __init__(self, num_heads, softmax_scale=None, attention_dropout=0.0, device=None, dtype=None):
+        super().__init__()
+        self.num_heads = num_heads
+        self.softmax_scale = softmax_scale
+        assert attention_dropout == 0
```
[the remark this answers is not in the document]
Comment (vchiley) [reply llm/src/flash_attention.py:37]: This entire file is a copy pasta from [hazy research](https://github.com/HazyResearch/flash-attention/blob/main/flash_attn/flash_attention.py); then modify it to use the triton kernel instead of the cuda flash attn kernel.
I kept it to keep the module signature the same, but to your point, I should probably just remove it and move the assertion higher in the stack.

File: llm/src/mosaic_gpt.py [hunk 11]
```diff
@@ -86,10 +90,80 @@ def __init__(self, cfg: DictConfig, device: Optional[str] = None):
...
+    alibi_bias = torch.arange(1 - seq_len, 1, dtype=dtype, device=device).view(1, 1, 1, seq_len)
+    if full:
+        # generate 1 x Heads x SeqLen x SeqLen alibi bias mask
+        # otherwise the mask is 1 x Heads x 1 x SeqLen (which is braodcasted up to the approproate size)
+        alibi_bias = alibi_bias - torch.arange(1 - seq_len, 1, dtype=dtype, device=device).view(1, 1, seq_len, 1)
+        alibi_bias.abs_().mul_(-1)  # since we're using causal flag, this isn't really needed, but why not include it
+
+    m = torch.arange(1, n_heads + 1, dtype=dtype, device=device)
+    m.mul_(alibi_bias_max / n_heads)
+    alibi_bias = alibi_bias * (1. / (2 ** m.view(1, n_heads, 1, 1)))
```
Comment (growlix) [inline llm/src/mosaic_gpt.py:165]: I didn't spend enough time doing the mental computations to figure out how this obtains the same values for _m_ as in the Composer implementation/paper, but your results look good so I'm not too concerned.

**Review by growlix — APPROVED** [review]
Great work! Two minor observations. Otherwise I endorse Alex and Abhi's reviews.

File: llm/src/flash_attn_triton.py (hunk 8 above)
Comment (dakinggg) [reply llm/src/flash_attn_triton.py:4]: perhaps we can link to the specific commit it is up to date with, so we don't lose track of that.

Comment (dakinggg) [reply llm/src/flash_attn_triton.py:4]: It would also be good to call out with comments anywhere that we have made changes. Not sure what the longer term plan is for keeping this in sync (or not) with hazy research.

File: llm/src/mosaic_gpt.py (hunk 11 above)
Comment (vchiley) [reply llm/src/mosaic_gpt.py:177]: Last sentence of 2nd paragraph of section 3 of [the alibi paper](https://arxiv.org/abs/2108.12409) goes over the details. This is my interpretation and convergence works well.

File: llm/src/flash_attn_triton.py (hunk 8 above)
Comment (vchiley) [reply llm/src/flash_attn_triton.py:4]: I symlinked to the BERT version of this file (but git doesn't respect symlinks)
changes can be see in this PRs changes to `bert/src/flash_attn_triton.py`
